app/particle/particle_simulation.py

```diff
@@ -152,29 +180,35 @@ def timestep():
 
 def change_num_active_workers():
     params.num_active_workers = params.new_num_active_workers
-    print("In change num_active_workers "+str(params.rank)+" "+str(params.num_active_workers))
     if params.rank is 0:
-        if params.new_num_active_workers < 1 or params.new_num_active_workers > params.num_threads - 1:
-            util.debug("Invalid number of active workers requested: " + params.new_num_active_workers)
-
         new_distribution = {}
         for i in range(1, params.num_threads):
             new_distribution[i] = set()
-
         for partition_id, partition in params.partitions.items():
             for particle in partition.particles:
                 new_thread = util.determine_particle_thread_num(particle.position[0])
                 particle.thread_num = new_thread
                 new_distribution[new_thread].add(particle)
-
         for i in range(1, params.num_threads):
-            print("sending "+str(i))
             params.partitions[i].particles = new_distribution[i]
             params.comm.send(new_distribution[i], dest = i, tag = 11)
     else:
-        print("receiving "+str(params.rank))
         params.partitions[params.rank].receive_new_particles()
-        #print("received "+str(params.rank))
+
+def change_num_particles():
+    if params.num_particles < params.new_num_particles:
+        if params.rank is 0:
+            for i in range(params.num_particles, params.new_num_particles):
+                add_one_particle(i)
+            for i in range(1, params.num_threads):
+                params.comm.send(params.partitions[i].particles, dest = i, tag = 11)
+        else:
+            params.partitions[params.rank].receive_new_particles()
+    elif params.num_particles > params.new_num_particles:
+        # note that params.new_num_particles == params.num_particles-1 is always true,
+        # and that subtract_one_particle() is implemented to only remove 1 particle
+        subtract_one_particle(params.new_num_particles)
```

Comment:
right now, the simulation has KEY_UP to add 1 particle (with a max limit of 200 particles) and KEY_DOWN to subtract 1 particle (with a min limit of 1 particle). so i thought that i should implement `subtract_one_particle` rather than `subtract_particles`, but i can implement `subtract_particles` if needed to!